### PR TITLE
[SR-3289] BE migration for backend disk balance

### PR DESCRIPTION
**Pull request by wyb**
1. Support tablet local migration for backend disk balance.
2. Improve backend tablet migration process.
3. Improve migration lock, and migration will not cause write fail.
4. Change path_hash type to int64, which is consistent with thrift api and fe.

### CODE REVIEW

File: be/src/agent/task_worker_pool.cpp [hunk 1]
```diff
@@ -896,16 +890,54 @@ void* TaskWorkerPool::_clone_worker_thread_callback(void* arg_this) {
...
+                }
+            }
         } else {
-            LOG(INFO) << "clone success, set tablet infos."
-                      << "signature:" << agent_task_req.signature;
-            finish_task_request.__set_finish_tablet_infos(tablet_infos);
+            EngineCloneTask engine_task(ExecEnv::GetInstance()->tablet_meta_mem_tracker(), clone_req,
+                                        worker_pool_this->_master_info, agent_task_req.signature, &error_msgs,
+                                        &tablet_infos, &status);
+            worker_pool_this->_env->storage_engine()->execute_task(&engine_task);
```
Comment (sduzh) [inline be/src/agent/task_worker_pool.cpp:928]: Check the result

**Review by sduzh — DISMISSED** [review]

File: be/src/storage/push_handler.cpp [hunk 2]
```diff
@@ -92,6 +92,10 @@ OLAPStatus PushHandler::_do_streaming_ingestion(TabletSharedPtr tablet, const TP
     if (!base_migration_rlock.owns_lock()) {
         return OLAP_ERR_RWLOCK_ERROR;
     }
+    if (Tablet::check_migrate(tablet)) {
```
Comment (chaoyli) [inline be/src/storage/push_handler.cpp:95]: Does migration is prohibited during migration?

File: be/src/storage/vectorized/delta_writer.cpp [hunk 3]
```diff
@@ -93,14 +93,29 @@ Status DeltaWriter::init() {
...
-            LOG(WARNING) << ss.str();
-            return Status::InternalError(ss.str());
+
+    // The tablet may have been migrated during delta writer init,
+    // and the latest tablet needs to be obtained when loading.
+    // Here, the while loop checks whether the obtained tablet has changed
+    // to get the latest tablet.
+    while (true) {
+        std::shared_lock base_migration_rlock(_tablet->get_migration_lock());
+        TabletSharedPtr new_tablet;
```
Comment (chaoyli) [inline be/src/storage/vectorized/delta_writer.cpp:103]: Does migration is prohibited during migration?

File: be/src/storage/task/engine_storage_migration_task.cpp [hunk 4]
```diff
@@ -161,57 +144,124 @@ OLAPStatus EngineStorageMigrationTask::_storage_medium_migrate(TTabletId tablet_
...
+            if (end_version != new_end_version) {
+                LOG(WARNING) << "copied version=" << end_version
+                             << " is different from new version=" << new_end_version;
+                need_remove_new_path = true;
+                res = OLAP_ERR_VERSION_NOT_EXIST;
+                break;
+            }
+
+            // generate new header
+            _generate_new_header(_dest_store, shard, tablet, consistent_rowsets, new_tablet_meta);
```
Comment (chaoyli) [inline be/src/storage/task/engine_storage_migration_task.cpp:221]: Does primary key table support this feature?

Comment (wyb) [reply be/src/storage/task/engine_storage_migration_task.cpp:221]: Primary key tablet migration is disabled and will be supported in the feature

**Review by chaoyli — DISMISSED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**